### PR TITLE
Publish CRD OpenAPI Documentation

### PR DESCRIPTION
**Pull request by roycaihw**
ref: https://github.com/kubernetes/enhancements/issues/692

### CODE REVIEW

**Comment by npentrel** [comment]
Hi @roycaihw, what is the status of this PR? We need all docs PRs to be ready to review by March 11th at the very latest. Thanks!

**Comment by npentrel** [comment]
Hi @roycaihw, is this ready for review? We need all docs PRs to be ready to review by March 11th at the very latest. Thanks!

**Comment by roycaihw** [comment]
the documentation is ready for review. Please take a look

/cc @liggitt @mbohlool @sttts

File: content/en/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions.md [hunk 1]
```diff
@@ -347,6 +347,31 @@ kubectl create -f my-crontab.yaml
...
+```
+--feature-gates=CustomResourcePublishOpenAPI=true
+```
+
+Custom resource validation schema will be converted to OpenAPI v2 schema, and
+show up in `definitions` and `paths` fields in the OpenAPI spec.
+Additional modification is applied during the conversion to keep backwards compatiblity with
+kubectl in previous 1.13 version. The conversion won't modify the validation schema defined in CRD,
+and therefore won't affect [validation](#validation) in API server. See
+[conversion implementation](https://github.com/kubernetes/kubernetes/blob/83ff0f6c64485ffe491d5116596072f466be1bd5/staging/src/k8s.io/apiextensions-apiserver/pkg/controller/openapi/conversion.go#L42-L112)
```
Comment (liggitt) [inline content/en/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions.md]: we don't typically link docs to code or pull requests... can you summarize the modifications done and the compatibility analysis here?

Comment (roycaihw) [reply content/en/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions.md]: added

I started with listing the modifications made, but I went with link because I thought it's wordy and may distract user who don't care about the modification

File: content/en/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions.md [hunk 2]
```diff
@@ -347,6 +347,62 @@ kubectl create -f my-crontab.yaml
 crontab "my-new-cron-object" created
 ```
 
+### Publish Validation Schema in OpenAPI v2
+
+{{< feature-state state="alpha" for_kubernetes_version="1.14" >}}
+
+Starting with Kubernetes 1.14, [custom resource validation schema](#validation) can be published as part
+of [OpenAPI v2 spec](docs/concepts/overview/kubernetes-api.md#openapi-and-swagger-definitions) from
+Kubernetes API server.
```
Comment (sttts) [inline content/en/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions.md]: some word about alpha?

File: content/en/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions.md [hunk 3]
```diff
@@ -347,6 +347,62 @@ kubectl create -f my-crontab.yaml
...
+  - the fields `type` and `properties` are removed
+- For a schema of `null` type
+  - the field `type` is removed
+- For a schema of `array` type
+  - if the schema doesn't have exact one item, the fields `type` and `items` are
+    removed
+- For a schema with no type specified
+  - the field `properties` is removed
+
+3. The following fields are removed as they aren't supported by gnostic
```
Comment (sttts) [inline content/en/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions.md]: s/gnostic/the OpenAPI protobuf implementation/

File: content/en/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions.md [hunk 4]
```diff
@@ -347,6 +347,62 @@ kubectl create -f my-crontab.yaml
...
+```
+
+Custom resource validation schema will be converted to OpenAPI v2 schema, and
+show up in `definitions` and `paths` fields in the OpenAPI spec.
+The following additional modifications are applied during the conversion to keep backwards compatiblity with
+kubectl in previous 1.13 version. These modifications prevent kubectl from being over-strict and rejecting
+valid openapi schemas that it doesn't understand. The conversion won't modify the validation schema defined in CRD,
+and therefore won't affect [validation](#validation) in API server.
+
+1. The following fields are removed as they aren't supported by OpenAPI v2
```
Comment (sttts) [inline content/en/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions.md]: we should maybe say that in future versions OpenAPI v3 will be used without these restrictions.

File: content/en/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions.md (hunk 2 above)
Comment (roycaihw) [reply content/en/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions.md]: added words about "Alpha in 1.14" in the paragraph below

File: content/en/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions.md [hunk 5]
```diff
@@ -347,6 +347,62 @@ kubectl create -f my-crontab.yaml
...
+### Publish Validation Schema in OpenAPI v2
+
+{{< feature-state state="alpha" for_kubernetes_version="1.14" >}}
+
+Starting with Kubernetes 1.14, [custom resource validation schema](#validation) can be published as part
+of [OpenAPI v2 spec](docs/concepts/overview/kubernetes-api.md#openapi-and-swagger-definitions) from
+Kubernetes API server.
+
+The published schema will be consumed by [kubectl](/docs/reference/kubectl/overview) to perform client-side
+validation on custom resources, and can be consumed for other purpose. The feature is Alpha in 1.14 and disabled by default.
```
Comment (sttts) [inline content/en/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions.md]: `kubectl explain` is missing.

File: content/en/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions.md [hunk 6]
```diff
@@ -347,6 +347,62 @@ kubectl create -f my-crontab.yaml
...
+You can enable the feature using the `CustomResourcePublishOpenAPI` feature gate on the
+[kube-apiserver](/docs/admin/kube-apiserver):
+
+```
+--feature-gates=CustomResourcePublishOpenAPI=true
+```
+
+Custom resource validation schema will be converted to OpenAPI v2 schema, and
+show up in `definitions` and `paths` fields in the OpenAPI spec.
```
Comment (sttts) [inline content/en/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions.md]: would add the OpenAPI v2 spec link from above here as well.

**Comment by roycaihw** [comment]
/assign @cody-clark 
for approval

File: content/en/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions.md [hunk 7]
```diff
@@ -347,6 +347,63 @@ kubectl create -f my-crontab.yaml
...
+and can be consumed for other purpose. The feature is Alpha in 1.14 and disabled by default.
+You can enable the feature using the `CustomResourcePublishOpenAPI` feature gate on the
+[kube-apiserver](/docs/admin/kube-apiserver):
+
+```
+--feature-gates=CustomResourcePublishOpenAPI=true
+```
+
+Custom resource validation schema will be converted to OpenAPI v2 schema, and
+show up in `definitions` and `paths` fields in the [OpenAPI v2 spec](docs/concepts/overview/kubernetes-api.md#openapi-and-swagger-definitions).
```
Comment (cody-clark) [inline content/en/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions.md]: Delete .md in link --> /docs/concepts/overview/kubernetes-api/#openapi-and-swagger-definitions

File: content/en/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions.md [hunk 8]
```diff
@@ -347,6 +347,63 @@ kubectl create -f my-crontab.yaml
 crontab "my-new-cron-object" created
 ```
 
+### Publish Validation Schema in OpenAPI v2
+
+{{< feature-state state="alpha" for_kubernetes_version="1.14" >}}
+
+Starting with Kubernetes 1.14, [custom resource validation schema](#validation) can be published as part
+of [OpenAPI v2 spec](docs/concepts/overview/kubernetes-api.md#openapi-and-swagger-definitions) from
```
Comment (cody-clark) [inline content/en/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions.md]: Delete .md in link --> /docs/concepts/overview/kubernetes-api/#openapi-and-swagger-definitions

File: content/en/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions.md [hunk 9]
```diff
@@ -347,6 +347,63 @@ kubectl create -f my-crontab.yaml
...
+  - the fields `type` and `properties` are removed
+- For a schema of `null` type
+  - the field `type` is removed
+- For a schema of `array` type
+  - if the schema doesn't have exact one item, the fields `type` and `items` are
+    removed
+- For a schema with no type specified
+  - the field `properties` is removed
+
+3. The following fields are removed as they aren't supported by the OpenAPI protobuf implementation
```
Comment (cody-clark) [inline content/en/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions.md]: This section is hard to read or scan. I'll approve this once the nits and lists are fixed, but I'd like to come up with a plan to organize this info better after the release.

I threw [something](https://hackmd.io/7A5CkswoRIiIQ6f5jU_FLg) together to present the info better, but it doesn't [render](https://5c8af24fff7f0f00085a5e4f--kubernetes-io-master-staging.netlify.com/docs/test/#table-for-removed-fields) the best on k8s.io because the inline code backticks are the same color as the alternating table cells.

Thoughts? Any other ideas?

File: content/en/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions.md [hunk 10]
```diff
@@ -347,6 +347,63 @@ kubectl create -f my-crontab.yaml
...
+### Publish Validation Schema in OpenAPI v2
+
+{{< feature-state state="alpha" for_kubernetes_version="1.14" >}}
+
+Starting with Kubernetes 1.14, [custom resource validation schema](#validation) can be published as part
+of [OpenAPI v2 spec](docs/concepts/overview/kubernetes-api.md#openapi-and-swagger-definitions) from
+Kubernetes API server.
+
+The published schema will be consumed by [kubectl](/docs/reference/kubectl/overview) to perform client-side
+validation (`kubectl create` and `kubectl apply`) and schema explanation (`kubectl explain`) on custom resources,
```
[the remark this answers is not in the document]
Comment (roycaihw) [reply content/en/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions.md]: updated

I left the last part unchanged ("The published schema can be consumed for other purposes") because it's not limited to only be used by `kubectl`. For example we know that `server-side apply` is planning to use the schema in future

File: content/en/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions.md (hunk 9 above)
Comment (roycaihw) [reply content/en/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions.md]: @cody-clark thanks for the suggestion. The table looks much better!

I'm not super concerned about the color, as long as we can tell the difference between code and the rest part of the sentence (bold can do the trick)

I'm concerned about the expressiveness of the table. e.g
- For a schema with a `$ref` (the `$ref` field in the schema object is defined)
- For a schema of `null` type (the `type` field in the schema object is defined and the value is `null`)

reads differently from

|Schema with  |
|---|
|`$ref`   |
|`null`   |

I'll think more about how to use table to arrange this section

**Comment by cody-clark** [comment]
Final Preview: https://deploy-preview-12910--kubernetes-io-vnext-staging.netlify.com/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions/#publish-validation-schema-in-openapi-v2

Thanks for following up, @roycaihw! I'll reach out after the release with something hopefully more concrete. Cheers!

/lgtm
/approve

**Comment by roycaihw** [comment]
> /hold

did it too late. I just noticed that a few things look wrong in the preview:
1. the bullet indentation is wrong in the ordered list (one layer is compacted)
2. the feature state is `Kubernetes v1.13 alpha` but not v1.14

@cody-clark any idea on how to fix? is 2) expected?
